### ISSUE TITLE
Fringe indicator not shown upon loading buffer

### ISSUE THREAD

**Issue by cashweaver** [nobiot/org-transclusion#126]
## What did you expect to happen?

I expect the fringe I have configured (see below) to appear after loading a file containing a transclusion. The transcluded text in the image below begins with "This is a new thing" and ends with "#+end_quote".

## What actually happened?

The transclusion was unadorned (see below) in the fringe and looked as if it were any other bit of text in the file. I was able to see the fringe indicator only after removing (`org-transclusion-remove`) and adding (`org-transclusion-add`) it again.

1. ![2022-01-17-20-57-44_screenshot](https://user-images.githubusercontent.com/56796112/149873516-7de65788-7fa8-408c-8572-a0db6ecb11a1.png)
2. Move `(point)` over the transclusion.
3. `M-x org-transclusion-remove`
3. `M-x org-transclusion-add`
4. ![2022-01-17-20-52-44_screenshot](https://user-images.githubusercontent.com/56796112/149873215-e8b366b7-ae18-4b18-b81e-117b49994934.png)
## Relevant config

I am running Doom Emacs with the following config:

```elisp
;; packages.el
(package! org-transclusion)

;; config.el
(use-package! org-transclusion
  :after org
  :config
  (setq
   left-fringe-width 14
   org-transclusion-exclude-elements '(property-drawer
                                       keyword)
   org-transclusion-extensions-loaded t
   org-transclusion-extensions '(org-transclusion-src-lines
                                 org-transclusion-font-lock
                                 org-transclusion-indent-mode))
  ;; Re-load extensions to activate `org-transclusion-indent-mode'.
  (org-transclusion-load-extensions-maybe t)
  (add-hook! 'org-mode-hook 'org-transclusion-mode)
  ;; (set-face-attribute
  ;;  'org-transclusion-fringe nil
  ;;  :foreground "white"
  ;;  :background nil)
  (define-fringe-bitmap 'org-transclusion-fringe-bitmap
    [17 34 68 136 68 34 17]
    nil nil 'center))
```

## Version

I'm using `org-transclusion` version `1.2.0`.

**Comment by nobiot** [comment]
Thanks for using Org-transclusion.

I cannot reproduce the issue on my end.
I  tested with a  minimal set up with only `org-transclusion` and relevant part of your conifg. See the image and code below.

![Screenshot from 2022-01-18 11-53-08](https://user-images.githubusercontent.com/12507865/149925487-88388945-5c28-4866-8d2e-f2cb3dc0b577.png)

I don't use Doom so don't know exactly what `use-package!`, `add-hook!` do, but I think it's safe to say your issue is caused by the loading sequence. Ensure to load `org-transclusion` before `define-fringe-bitmap`. I suspect it may be done after your custom bitmap definition and thus `org-transclusion` overrides it.

---

Script  for minimal setup I used:

As noted in the comment of the script, I removed my own config directory and files, and did `emacs -l load-org-trasnclusion.el (the filename of this  script).

``` elisp
;; -*- lexical-binding: t; -*-

;; Test after removing config files
;; In terminal:
;;    $ emacs -l load-org-trasclusion.el

(package-refresh-contents)
(package-install 'org-transclusion)
(add-hook 'org-mode-hook #'org-transclusion-mode)
(with-eval-after-load 'org-transclusion
  (define-fringe-bitmap 'org-transclusion-fringe-bitmap
    [17 34 68 136 68 34 17]
    nil nil 'center))
```

**Comment by cashweaver** [comment]
![2022-01-18-06-12-59_screenshot](https://user-images.githubusercontent.com/56796112/149953595-2529fb7c-d6d3-45b7-a1f2-9fa8deae47f5.png)

The fringe bitmap which indicates the transclusion source _is_ visible when I first load the file.

**Comment by nobiot** [comment]
Yes. The source one uses a different fringe bitmap and you don't change it, so I don't think it is related.

My guess is that before your custom bitmap appears, the standard one is actually used. It's just not visible against the dark background of the theme you use.

**Comment by nobiot** [comment]
I think you need to move this hook definition elsewhere, outside the org-transclusion part of use-package! or add it to :init part.

> (add-hook! 'org-mode-hook 'org-transclusion-mode)

Notice in my config add-hook is not included in eval-after-load?

I think you want to activate org-transclusion in every org file you visit. Does it actually work in the very first org visit immediately after you launch Emacs?

The way you put add-hook! looks a bit odd to me. The hook would get added only after the org-transclusion has been loaded. first time. But this does not happen until you load org. So the sequence seems to me

You visit an Org file -> Emacs loads Org -> Emacs loads org-transclusion -> Emacs add-hook -> Control returns to you. It’s just adding a hook, not activating org-transclusion-mode. You then need manually do it. The second org file onwards, the hook is active so transclusion-mode turns on automatically. 

I can’t tell why bitmap gets does not get changed in the first call to org-transclusion.

**Comment by nobiot** [comment]
Okay, I can reproduce your issue on Doom but I cannot on my vanilla Emacs.
I now suspect there might something different between the two in the way Emacs loads `org-indent-mode and the extension `org-transclusion-indent-mode`.... I cannot be sure.

**Comment by rtrppl** [comment]
I may have a similar issue with plain Emacs (not using Doom). I was able to change the color for the fringe of the source, but not for `org-transclusion-fringe`. Is there a setting that could override `org-transclusion-fringe`? My apologies, if this is an unrelated issue. I am using the "zenburn" theme.

![Bildschirmfoto 2022-01-22 um 22 30 27](https://user-images.githubusercontent.com/63909884/150656273-5fa1f5f7-a864-4330-b2cd-fa8c1e6ce9dc.png)

![Bildschirmfoto 2022-01-22 um 22 34 05](https://user-images.githubusercontent.com/63909884/150656284-d413c0a2-3434-48b8-abf4-702740097beb.png)

I love the idea of this for writing a non-fiction book!

**Comment by rtrppl** [comment]
Quick follow-up: It works for me if I turn on the "indent-mode" of the `org-transclusion-extensions`.

<img width="1471" alt="Bildschirmfoto 2022-01-23 um 09 29 15" src="https://user-images.githubusercontent.com/63909884/150670691-7b0705db-a5db-40a9-88aa-3e293fc8101e.png">

**Comment by nobiot** [comment]
Thank you. I think your issue is unrelated but old like to investigate further. Do you use the built-in org-indent-mode? It does not look like you do from the screen shots. it’s a little strange that you would have to turn on this extension…

**Comment by rtrppl** [comment]
I actually do use it. There was just no headline in the screenshots above. After adding one, I do see another weird thing: if I open the the transclusion, the indent in the source is lost.

<img width="1471" alt="Bildschirmfoto 2022-01-23 um 09 46 48" src="https://user-images.githubusercontent.com/63909884/150671201-bba107ab-4d67-44b0-99c4-d86caded0dd5.png">

**Comment by nobiot** [comment]
I see; that explains. Thank you. org-indent-mode rather aggressively inserts the indentation, and I have struggled to support it (it was not supported originally and later added it only as an extension). 

At the moment, unfortunately losing the source’s indentation is an expected behavior and limitation. I will need to add it to the user manual as a known limitation…

**Comment by rtrppl** [comment]
Thanks for the response. It does not bother me but `org-transclusion` also seems to introduce a first-line indentation. 

![Bildschirmfoto 2022-01-23 um 13 43 08](https://user-images.githubusercontent.com/63909884/150678953-db510ae6-0cbe-424c-ada2-c62e3b1adf7b.png)

**Comment by jlcolbert** [comment]
I have a Doom solution:

```
(use-package! org-transclusion
  ;; :defer t
  :after org
  :init   ; you technically don't need to put map! in this :init section, you could put it outside the use-package! declaration
  (map!
    :map "<f12>" #'org-transclusion-add ; you don't need global-map with the map! macro
    :leader
    :prefix "n"
    :desc "Org Transclusion Mode" "t" #'org-transclusion-mode)
  :config
  (add-to-list 'org-transclusion-extensions 'org-transclusion-indent-mode)
  (custom-set-faces! `(org-transclusion-fringe ; the backwards tick as opposed to apostrophe is *crucial*
                        :foreground ,(doom-color 'green)
                        :background ,(doom-color 'green))))
```

To set the fringe colors, you need to use [`custom-set-faces!`](https://github.com/hlissner/doom-emacs/blob/aed2972d7400834210759727117c50de34826db9/core/autoload/themes.el#L46). To make sure it fits well with your theme, use `doom-color` plus the name of the color you want. I cycled through light and dark themes, and the green changed appropriately for each one. I imagine it would work fine with whatever other colors you choose.

You don't need to explicitly require the indent extension after adding it to the list. At least, things still work fine when I don't. And you can either explicitly defer the package until you call it, or you can have it load after org. Both worked fine.

When I tried using a hook to load in every org buffer, I would have to disable the mode and enable it again in order for it to work. Having the mode off until called works fine for my workflow.

**Comment by nobiot** [comment]
@jlcolbert Thank you for sharing this!

**Comment by jlcolbert** [comment]
@nobiot No problem! Here's my full configuration with the keybindings:

```
(use-package! org-transclusion
  :defer t
  :config
  (add-to-list 'org-transclusion-extensions 'org-transclusion-indent-mode)
  (custom-set-faces! `(org-transclusion-fringe
                        :foreground ,(doom-color 'green)
                        :background ,(doom-color 'green)))
  (map! :map org-transclusion-map ; if you put :after org-transclusion in this map declaration, you don't need to put it in use-package!
    :prefix "C-c"
    [backtab] #'org-cycle ; Doom has `org-toggle-inline-images' bound to C-c TAB, but has `org-cycle' bound to global TAB in evil normal state
    "D" #'org-transclusion-demote-subtree
    "O" #'org-transclusion-move-to-source
    "P" #'org-transclusion-promote-subtree
    "d" #'org-transclusion-remove
    "e" #'org-transclusion-live-sync-start
    "g" #'org-transclusion-refresh
    "o" #'org-transclusion-open-source))

(map! "<f12>" #'org-transclusion-mode
      :leader ; in Doom the leader key is SPC (if you have evil enabled)
      (:prefix "n"
        (:prefix ("T" . "transclusion")
          :desc "add" "a" #'org-transclusion-add
          :desc "add all" "A" #'org-transclusion-add-all
          :desc "make from link" "m" #'org-transclusion-make-from-link)))
```

I will continue to play around with the bindings to see if I can get them all in one declaration, but I don't claim to know how a lot of this works beyond reading the docs and playing around until something sticks.

**Comment by ahmed-shariff** [comment]
I also have the same issue, but I don't use org-indent.  I was playing around to see what was happening:

The `org-transclusion-content-insert` function seems to be setting the `line-prefix` and `warp-prefix` for the corresponding region as text-properties for the transclusion region. But uses the overlays to modify that for the source region. After transcluding, when checking the text-properties of the tranclusion region, `line-prefix` and `warp-prefix` are nil. I don't fully understand how font lock and text properties work, but I am assuming this is why the source buffer has the fringe and the tranclusion buffer doesn't. With the cursor on the tranclusion region, if I do the following: 

```emacs-lisp
(overlay-put (text-clone-make-overlay (marker-position (get-text-property (point) 'org-transclusion-beg-mkr)) (marker-position (get-text-property (point) 'org-transclusion-end-mkr)) (current-buffer)) 'wrap-prefix (org-transclusion-propertize-transclusion))
```
(which is using overlay to set the  `line-prefix` and `warp-prefix` by extracting other values in the text properties)
I get the fringe in the buffer.

As a workaround, I am doing the following:
```emacs-lisp
  (defun org-transclusion-content-insert-add-overlay (beg end)
    "Add fringe after transclusion."
    (overlay-put (text-clone-make-overlay beg end (current-buffer))
                 'line-prefix
                 (org-transclusion-propertize-transclusion))
    (overlay-put (text-clone-make-overlay beg end (current-buffer))
                 'wrap-prefix
                 (org-transclusion-propertize-transclusion)))

  (add-hook 'org-transclusion-after-add-functions #'org-transclusion-content-insert-add-overlay)
```

**Comment by ahmed-shariff** [comment]
@nobiot is it worth refactoring so that adding the fringe is done with the overlay functions instead of text-properties?

**Comment by nobiot** [comment]
@ahmed-shariff Thanks for the workaround and sharing. Much appreciated.

I have to ask what is your "same issue" is.

This thread has been dormant for a while, and I can see there were multiple issues (albeit being similar) discussed above. If you are so inclined, I'd appreciate if you could open an new issue with a fresh description of what it is that your work around addresses.

Having said this, however, I am sorry to that I don't have much capacity to deep-dive into the issue. I imagine it would not be trivial. My gut is at the moment inclined to saying it is probably not worth refactoring to switch from text-props to overlays. An overlay is already added back recently  (so now we have both) with https://github.com/nobiot/org-transclusion/commit/bc24be23348a6849868b6c49b865e56a175b50c5:

![image](https://github.com/nobiot/org-transclusion/assets/12507865/3645ab92-2a94-4b63-9fa7-4c9edf750a63)

The green color is a face to the overlay, while the fringe's vertical line comes from the text-props. I think you could play with it to add the line- and wrap-prefix.

I had changed from overlays to text-properties a while before adding the overlay back. My notes do not really give me much as to why the removal (my bad), but I believe it is related to the behavior with `undo`. Text-props are part of the text, so they are in sync with the text when you undo -- you can undo `org-transclusion-add`. The behavior of having overlay only will be anti-intuitive especially when you use the live-sync function and undo together.

**Comment by nobiot** [comment]
This commit 4f9b8fa from around 2 years ago and before v1.0.0 already have text-property for yanking from the live-sync. Just reinforcing my gut feel that switching from text-prop to overlay would be unlikely to be trivial. Just FYI @ahmed-shariff

**Comment by ahmed-shariff** [comment]
Thank you for the quick reply. Apologize for necroing this thread, I'll open a new issue for this discussion.

**Comment by webloginwu** [comment]
> Okay, I can reproduce your issue on Doom but I cannot on my vanilla Emacs. I now suspect there might something different between the two in the way Emacs loads `org-indent-mode` and the extension `org-transclusion-indent-mode`.... I cannot be sure.
> 
> I'm sorry but it's really hard to support Doom users properly; it's too complex for me.

I just provide infomation:
My config is simple:

```
(use-package org-transclusion
  :init
  (org-transclusion-mode 0))
(setq org-startup-indented t)

```
I have two laptop:
on mac mini ,      I can        see Fringe indicator after I run M-x org-transclusion-add, it is ok.
on mac book air, I can not see Fringe indicator after I run M-x org-transclusion-add
work around:
I toggle off org-indent-mode on on mac book air, then I can see Fringe indicator after I run M-x org-transclusion-add

**Comment by dustinfarris** [comment]
I hit this on Doom Emacs today and traced it to a root cause that I think explains both the original report and why it reproduces on Doom but not on vanilla `emacs -Q`.

## Root cause: `org-indent-mode`'s async init restamps `line-prefix` after the render

The fringe indicator lives in the `line-prefix`/`wrap-prefix` text properties. `org-indent-mode` writes those same properties — and it initializes **asynchronously** (`org-indent-initialize-agent`, an idle timer). So on `find-file` the ordering is:

1. `org-mode-hook` runs → `org-transclusion-mode` renders the transclusions and puts the fringe in `line-prefix` (fringe is briefly visible),
2. moments later org-indent's agent runs `org-indent-add-properties` over the whole buffer, replacing `line-prefix`/`wrap-prefix` wholesale → fringe gone.

This isn't Doom-specific — it's `org-indent-mode`-specific. Doom enables `org-indent-mode` in org buffers by default, which is why it reproduces there and not in a bare session. (Note @webloginwu's config above has `(setq org-startup-indented t)` — same trigger.) A minimal repro from `emacs -Q` is just:

```elisp
(add-hook 'org-mode-hook #'org-indent-mode)
(add-hook 'org-mode-hook #'org-transclusion-mode)
;; open a file containing a #+transclude: directive → fringe flashes, then disappears
```

The `org-transclusion-indent-mode` extension doesn't cover this case: its destination-buffer repair (`org-transclusion-indent--add-properties-and-fringes`) only runs on `org-transclusion-after-add-functions` — i.e. at add time, *before* the agent's stamp — and `org-transclusion-indent--reapply-all-fringes` only walks **source** overlays, so the destination buffer is never re-repaired. That also matches the observed remedy: `org-transclusion-remove` + `org-transclusion-add` re-runs the add-time repair after org-indent has settled, so the fringe sticks.

## A second, related bug: `org-transclusion-prefix-has-fringe-p` never inspects position 0

While debugging, I found the detection predicate has a blind spot:

```elisp
(let ((pos 0))
  (catch 'found
    (while (setq pos (next-single-property-change pos 'display prefix))
      ...))
```

`next-single-property-change` returns the next position where the property *changes* — it never yields position 0 itself, and for a prefix string whose `display` property spans the whole string it returns `nil` immediately. So a fringe-**only** prefix (the common case when `org-indent-mode` is off, and also headline lines whose org-indent prefix is the empty string) is reported as *fringeless*. Callers that guard on this predicate then mis-detect state; anything that logs or conditionally re-applies based on it gets false negatives.

The rewritten predicate in #285 scans every position starting from 0, so it would fix this implicitly — but I don't think it's been called out as a bug anywhere, so noting it here in case that refactor gets rebased.

## Workaround

Re-appending the fringe after every org-indent stamp fixes it durably (open, revert, save cycles all survive):

```elisp
(defun my/org-transclusion-keep-fringes (beg end &rest _)
  "Re-append transclusion fringes after `org-indent-add-properties'."
  (when (bound-and-true-p org-transclusion-mode)
    (save-excursion
      (goto-char beg)
      (let (match)
        (while (and (< (point) end)
                    (setq match (text-property-search-forward
                                 'org-transclusion-type)))
          (when (< (prop-match-beginning match) end)
            (org-transclusion-add-fringe-to-region
             (current-buffer) (prop-match-beginning match)
             (min (prop-match-end match) end)
             'org-transclusion-fringe)))))))

(advice-add #'org-indent-add-properties :after #'my/org-transclusion-keep-fringes)
```

(The position-0 blind spot in `org-transclusion-prefix-has-fringe-p` is harmless here: `add-fringe-to-region` may re-append to a fringe-only prefix it fails to recognize, but each org-indent stamp replaces the prefix first, so nothing accumulates.)

A proper in-package fix might be for `org-transclusion-indent-mode` to install exactly this kind of `:after` advice (or hook) on `org-indent-add-properties` for destination buffers, alongside its existing source-overlay repair.

Tested on org-transclusion `feda2f0` (current main), Org 9.7, Emacs 30.2, macOS GUI.
